### PR TITLE
mrustc: init at 0.9

### PR DESCRIPTION
**Pull request by r-burns**
This is rebased from the excellent work done by @ProgVal in https://github.com/NixOS/nixpkgs/pull/85542, with the hopes that we can merge the underlying mrustc expressions and iterate on them while we sort out the bootstrapping logic.

Functional changes:
* Needs compiler-rt patch for glibc ≥ 2.31

Cleanups:
* Add pre/post hooks for all steps
* Reuse src from main mrustc derivation
* Use strictDeps
* Override make variables with makeflags rather than patches + env vars

The mrustc-bootstrap builds rustc and cargo using mrustc. This is the main
motivator for mrustc and is very finicky so it's important that we
build this regularly and keep it green.

@ProgVal says:
> The initial build of Rust 1.29 (by mrustc) requires 2.5GB in /build
  and 15GB of RAM and takes about 30 minutes on a modern x86 machine.

Can I tell hydra that this build will need a lot of RAM?
Compilation might take several hours on less beefy machines;
I hope Hydra doesn't timeout on this.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Please squash all package related commits together.

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    llvm_7 libffi zlib libxml2
+    # for cargo
+    openssl curl
+  ];
+
+  makeFlags = [
+    # Use shared mrustc/minicargo/llvm instead of rebuilding them
+    "MRUSTC=${mrustc}/bin/mrustc"
+    "MINICARGO=${mrustc-minicargo}/bin/minicargo"
+    "LLVM_CONFIG=${llvm_7}/bin/llvm-config"
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/mrustc/bootstrap.nix:80]: I think we should use buildPackages.llvm_7 here to not break cross compiling.

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    for crate in regex regex-0.2.11 curl-sys
+    do
+      echo "building $crate"
+      minicargo ${rustcDir}/src/vendor/$crate \
+        --vendor-dir ${rustcDir}/src/vendor \
+        --output-dir output/cargo-build -L output/
+    done
+
+    echo minicargo.mk: rustc
+    make -f minicargo.mk "''${flagsArray[@]}" output/rustc
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/mrustc/bootstrap.nix:108]: Is this using NIX_BUILD_CORES?

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    runHook postCheck
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin/ $out/lib/
+    cp run_rustc/output/prefix/bin/cargo $out/bin/cargo
+    cp run_rustc/output/prefix/bin/rustc_binary $out/bin/rustc
+
+    cp -r run_rustc/output/prefix/lib/* $out/lib/
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/mrustc/bootstrap.nix:134]: Maybe we should move this to a separate output.

**Comment by r-burns** [comment]
> I've kept my fixes/cleanups in separate commits only for ease of review, and plan to squash before merging in accordance with CONTRIBUTING guidelines.

@SuperSandro2000 Is this all right with you? As a reviewer I sometimes like to see separation of concerns in commits, but I can just squash if it's not actually helpful.

File: pkgs/development/compilers/mrustc/bootstrap.nix (hunk 2 above)
Comment (ProgVal) [reply pkgs/development/compilers/mrustc/bootstrap.nix:108]: No, minicargo is single-threaded

File: pkgs/development/compilers/mrustc/bootstrap.nix (hunk 3 above)
Comment (r-burns) [reply pkgs/development/compilers/mrustc/bootstrap.nix:134]: It's definitely something we can look into, but I think for now I would prefer to just keep the bootstrap installation as simple as possible.  I expect the final user-facing rustc derivations will disallow references to these bootstrapping stages anyway, so it shouldn't make any difference whether this uses split outputs.

File: pkgs/development/compilers/mrustc/bootstrap.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/compilers/mrustc/bootstrap.nix:80]: Actually not. If I can finish https://github.com/NixOS/nixpkgs/pull/111487, this should be one of those `$dev/bin` config things that is actutally for the build platform. See #51176 it's unfortunate.

**Comment by SuperSandro2000** [comment]
> Is this all right with you?

I bet 10 bucks on that I will probably forget it.

> As a reviewer I sometimes like to see separation of concerns in commits, but I can just squash if it's not actually helpful.

If you don't rebase with master on force pushes I can see from the force push diff what you did. I am also using the mark-files-as-read-feature. Either works for me but keeping review comments around takes another force push before it can be merged.

**Comment by r-burns** [comment]
> I bet 10 bucks on that I will probably forget it.

Haha, fair enough :)
Squashed.

**Comment by r-burns** [comment]
@ofborg build mrustc-minicargo

**Comment by ProgVal** [comment]
I started updating the nix expression to work with the current mrustc master, which supports rust 1.39: https://github.com/ProgVal/nixpkgs/commit/07b59ad2e72d537624d7f53539942236313f6c38

But I still can't use it to build even build rust 1.29 (`error: cannot satisfy dependencies so `std` only shows up once`) and can't find how to fix it. Feel free to pick up where I stopped

**Comment by grahamc** [comment]
Does the build use many CPU cores if they're available, or is it a memory-bound operation?

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    # for cargo
+    openssl curl
+  ];
+
+  makeFlags = [
+    # Use shared mrustc/minicargo/llvm instead of rebuilding them
+    "MRUSTC=${mrustc}/bin/mrustc"
+    "MINICARGO=${mrustc-minicargo}/bin/minicargo"
+    "LLVM_CONFIG=${llvm_7}/bin/llvm-config"
+    "RUSTC_TARGET=${stdenv.buildPlatform.config}"
```
Comment (alyssais) [inline pkgs/development/compilers/mrustc/bootstrap.nix:81]:
```suggestion
    "RUSTC_TARGET=${rust.toRustTarget stdenv.targetPlatform}"
```

File: pkgs/development/compilers/mrustc/bootstrap.nix [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin/ $out/lib/
+    cp run_rustc/output/prefix/bin/cargo $out/bin/cargo
+    cp run_rustc/output/prefix/bin/rustc_binary $out/bin/rustc
+
+    cp -r run_rustc/output/prefix/lib/* $out/lib/
+    cp $out/lib/rustlib/${stdenv.buildPlatform.config}/lib/*.so $out/lib/
```
Comment (alyssais) [inline pkgs/development/compilers/mrustc/bootstrap.nix:135]:
```suggestion
    cp $out/lib/rustlib/${rust.toRustTarget stdenv.targetPlatform}/lib/*.so $out/lib/
```

**Review by alyssais — COMMENTED** [review]
LGTM otherwise.  I don't think we need to wait for the latest version to get an mrustc in Nixpkgs.

**Comment by ProgVal** [comment]
@grahamc It's partially parallel, but its bottleneck is by far the compilation of `rustc` (the main crate) by mrustc, which is not parallelizable

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113817` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>mrustc</li>
<li>mrustc-bootstrap</li>
<li>mrustc-minicargo</li>
</ul>
</details>

**Comment by r-burns** [comment]
@alyssais Thanks for the suggestions, updated.

**Comment by alyssais** [comment]
@OfBorg build mrustc-minicargo

**Comment by alyssais** [comment]
Now let's get an up to date one!

Feel free to request a review from me in any followups.

**Comment by alyssais** [comment]
@r-burns jsyk, in future when you take a commit like that, you can mark
yourself as a co-author on that commit using Co-authored-by, so that
GitHub and other tools will recognise your work on it.

https://docs.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

**Comment by r-burns** [comment]
Yay, it built! https://hydra.nixos.org/build/141770406#tabs-summary
Took about two hours, so I guess we don't need to set any special hydra rules for this.

**Pull request merged**